### DISCUSSION TITLE
Possibly better RxJS pipe typings?

### DISCUSSION THREAD

**Discussion by k-tten** [ReactiveX/rxjs#6671] [Ideas / Feature request]
**I know it's unlikely for someone to use 9+ operators in a pipe... but this is really bugging me...**

[`pipe.ts' typings`](https://github.com/ReactiveX/rxjs/blob/master/src/internal/util/pipe.ts#L4-L70) seems pretty huge, redundant, and unnecessary, and I want to help reduce it so:
- It's not huge, redundant, and unnecessary
- It can take 9+ operators and still produce the correct type

I made a simple prototype with this working in this [gist](https://gist.github.com/cursorsdottsx/04a3cd3d5a8ee1532f64e82094f434c8) (updated)

Currently, it takes in any amount of unary functions, and does the following:
- Validates if the unary functions are composable together (parameter and return types match)
- Spits out an unary function of the first one's parameter and the last one's return type
- If no arguments are present it spits out the identity function
- If only one argument is present it returns that argument

Thoughts, anyone?

Let me know if this can be possibly added.

Related issues: #4177

**Comment by kwonoj** [comment]
Moved to discussion as there's already filed issue regarding number of overload, and this is a suggestion.

**Comment by voliva** [comment]
There's a few challenges with this implementation though.

Although the typings for this pipe verify that everything is consistent and returns the correct input/output, it doesn't let typescript infer any of the parameters.

Take this example (it's essentially the one you have in gist, but I've removed the type anotation of the parameter for the second operator):

```ts
pipe(
  (_: string) => 0, // Error: Argument of type '(_: string) => number' is not assignable to 'never'
  (value) => String(value), // Error: Parameter 'value' implicitly has an 'any' type
  (_: string) => true,
);
```

I would expect in this case for TS to infer that `value` is of type number, and to not output any error.

This is specially relevant with rxjs, where you mostly expect everything to be inferred:

```ts
pipe(
  (value: number) => of(value),
  map(myValue => myValue.toFixed(3)), // myValue is inferred as number
  filter(myString => myString.includes('2')) // myString is inferred as string
);
```

I've tried many times to write the types of a pipe function that can do this, but I've never managed to get something working. It would be amazing to see a solution to this problem!

**Reply by k-tten** [reply]
Good point, I totally forgot about that. However I think it is still possible by modifying the parameter validator :)

Could you possibly link a gist of your attempts? I want to set aside some time to sit down and try a few things for a few hours.
